File: libretro-db/dat_converter.c
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

#include "rarchdb.h"
#include "db_parser.h"

#define MAX_TOKEN 256

static char *strndup_(const char *s, size_t n)
{
   char* buff = calloc(n, sizeof(char));

   if (!buff)
      return 0;

   strncpy(buff, s, n);
   return buff;
}

static struct rmsgpack_dom_value *get_map_value(
      const struct rmsgpack_dom_value *m, char* key)
{
   struct rmsgpack_dom_value v;

   v.type = RDT_STRING;
   v.string.len = strlen(key);
   v.string.buff = key;
   return rmsgpack_dom_value_map_value(m, &v);
}

static int load_string(int fd, struct rmsgpack_dom_value *out)
{
   char tok[MAX_TOKEN];
   ssize_t tok_size;

   if ((tok_size = get_token(fd, tok, MAX_TOKEN)) < 0)
      return tok_size;

   out->type = RDT_STRING;
   out->string.len = tok_size;
   out->string.buff = strndup_(tok, tok_size);
   return 0;
}

static int load_uint(int fd, struct rmsgpack_dom_value *out)
{
   char tok[MAX_TOKEN], *c;
   ssize_t tok_size;
   uint64_t value = 0;

   if ((tok_size = get_token(fd, tok, MAX_TOKEN)) < 0)
      return tok_size;

   for (c = tok; c < tok + tok_size; c++)
   {
      value *= 10;
      value += *c - '0';
   }

   out->type = RDT_UINT;
   out->uint_ = value;
   return 0;
}

static int load_bin(int fd, struct rmsgpack_dom_value *out)
{
   char tok[MAX_TOKEN];
   ssize_t tok_size;
   uint8_t h;
   uint8_t l;
   int i;

   if ((tok_size = get_token(fd, tok, MAX_TOKEN)) < 0)
      return tok_size;

   out->type = RDT_BINARY;
   out->binary.len = tok_size / 2;

   for (i = 0; i < tok_size; i += 2)
   {
      if (tok[i] <= '9')
         h = tok[i] - '0';
      else
         h = (tok[i] - 'A') + 10;
      if (tok[i+1] <= '9')
         l = tok[i+1] - '0';
      else
         l = (tok[i+1] - 'A') + 10;
      tok[i/2] = h * 16 + l;
   }

   out->binary.buff = malloc(out->binary.len);
   memcpy(out->binary.buff, tok, out->binary.len);
   return 0;
}

static int dat_value_provider(void *ctx, struct rmsgpack_dom_value *out)
{
   int rv, i;
   static const int field_count = 22;
   int fd = *((int*)ctx);
   char* key;

   out->type = RDT_MAP;
   out->map.len = field_count;
   out->map.items = calloc(field_count, sizeof(struct rmsgpack_dom_pair));

   if (find_token(fd, "game") < 0)
      return 1;

   for (i = 0; i < field_count; i++)
   {
      if ((rv = load_string(fd, &out->map.items[i].key)) < 0)
         goto failed;

      key = out->map.items[i].key.string.buff;

      if (strncmp(key, "name", sizeof("name")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "description", sizeof("description")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "users", sizeof("users")) == 0)
      {
         if ((rv = load_uint(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "releasemonth", sizeof("releasemonth")) == 0)
      {
         if ((rv = load_uint(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "releaseyear", sizeof("releaseyear")) == 0)
      {
         if ((rv = load_uint(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "rumble", sizeof("rumble")) == 0)
      {
         if ((rv = load_uint(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "analog", sizeof("analog")) == 0)
      {
         if ((rv = load_uint(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "serial", sizeof("serial")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "esrb_rating", sizeof("esrb_rating")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "elspa_rating", sizeof("elspa_rating")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "pegi_rating", sizeof("pegi_rating")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "cero_rating", sizeof("cero_rating")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "developer", sizeof("developer")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "publisher", sizeof("publisher")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "origin", sizeof("origin")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "rom", sizeof("rom")) == 0)
      {
         if (find_token(fd, "name") < 0)
            goto failed;

         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "size", sizeof("size")) == 0)
      {
         if ((rv = load_uint(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "sha1", sizeof("sha1")) == 0)
      {
         if ((rv = load_bin(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "crc", sizeof("crc")) == 0)
      {
         if ((rv = load_bin(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "md5", sizeof("md5")) == 0)
      {
         if ((rv = load_bin(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, "serial", sizeof("serial")) == 0)
      {
         if ((rv = load_string(fd, &out->map.items[i].value)) < 0)
            goto failed;
      }
      else if (strncmp(key, ")", sizeof(")")) == 0)
      {
         rmsgpack_dom_value_free(&out->map.items[i].key);
         out->map.len = i;
         printf("Couldn't find all fields for item\n");
         break;
      }
      else
      {
         rmsgpack_dom_value_free(&out->map.items[i].key);
         i--;
      }
   }
   printf("Inserting '%s' (%02X%02X%02X%02X)...\n",
         get_map_value(out, "name")->string.buff,
         (unsigned char)get_map_value(out, "crc")->binary.buff[0],
         (unsigned char)get_map_value(out, "crc")->binary.buff[1],
         (unsigned char)get_map_value(out, "crc")->binary.buff[2],
         (unsigned char)get_map_value(out, "crc")->binary.buff[3]
         );
   return 0;

failed:
   rmsgpack_dom_value_free(out);
   out->type = RDT_NULL;
   return rv;
}

int main(int argc, char **argv)
{
   int rv = 0;
   int src = -1;
   int dst = -1;

   if (argc != 3)
      printf("Usage: %s <dat file> <output file>\n", argv[0]);

   src = open(argv[1], O_RDONLY);

   if (src == -1)
   {
      printf("Could not open source file '%s': %s\n", argv[1], strerror(errno));
      rv = errno;
      goto clean;
   }

   dst = open(argv[2], O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);

   if (dst == -1)
   {
      printf("Could not open destination file '%s': %s\n", argv[1], strerror(errno));
      rv = errno;
      goto clean;
   }

   rv = rarchdb_create(dst, &dat_value_provider, &src);

clean:
   if (src != -1)
      close(src);
   if (dst != -1)
      close(dst);
   return rv;
}

File: libretro-db/lua/dat_converter.lua
local dat_obj = {}
local match_key = nil

local function dat_lexer(f, fname)
    local line, err = f:read("*l")
    local location = {line_no = 1, column = 1, fname = fname}
    return function()
        local tok = nil
        while not tok do
            if not line then
                return nil
            end
            pre_space, tok, line = string.match(line, "^(%s*)(..-)([()]*%s.*)")
            if tok and string.match(tok, "^\"") then
                tok, line = string.match(tok..line, "^\"([^\"]-)\"(.*)")
            elseif tok and string.match(tok, "^[()]") then
                line = tok:sub(2) .. line
                tok = tok:sub(1,1)
            end
            location.column = location.column  + #(pre_space or "")
            tok_loc = {
                line_no = location.line_no,
                column = location.column,
                fname = location.fname
            }
            if not line then
                line = f:read("*l")
                location.line_no = location.line_no + 1
                location.column = 1
            else
                location.column = location.column + #tok
            end
        end
        -- print(tok)
        return tok, tok_loc
    end
end

local function dat_parse_table(lexer, start_loc)
    local res = {}
    local state = "key"
    local key = nil
    for tok, loc in lexer do
        if state == "key" then
            if tok == ")" then
                return res
            elseif tok == "(" then
                error(string.format(
                    "%s:%d:%d: fatal error: Unexpected '(' instead of key",
                    loc.fname,
                    loc.line_no,
                    loc.column
                ))
            else
                key = tok
                state = "value"
            end
        else
            if tok == "(" then
                res[key] = dat_parse_table(lexer, loc)
            elseif tok == ")" then
                error(string.format(
                    "%s:%d:%d: fatal error: Unexpected ')' instead of value",
                    loc.fname,
                    loc.line_no,
                    loc.column
                ))
            else
                res[key] = tok
            end
            state = "key"
        end
    end
    error(string.format(
        "%s:%d:%d: fatal error: Missing ')' for '('",
        start_loc.fname,
        start_loc.line_no,
        start_loc.column
    ))
end

local function dat_parser(lexer)
    local res = {}
    local state = "key"
    local key = nil
    local skip = true
    for tok, loc in lexer do
        if state == "key" then
            if tok == "game" then
                skip = false
            end
            state = "value"
        else
            if tok == "(" then
                local v = dat_parse_table(lexer, loc)
                if not skip then
                    table.insert(res, v)
                    skip = true
                end
            else
                error(string.format(
                    "%s:%d:%d: fatal error: Expected '(' found '%s'",
                    loc.fname,
                    loc.line_no,
                    loc.column,
                    tok
                ))
            end
            state = "key"
        end
    end
    return res
end

local function unhex(s)
    if not s then return nil end
    return (s:gsub('..', function (c)
        return string.char(tonumber(c, 16))
    end))
end

local function get_match_key(mk, t)
    for p in string.gmatch(mk, "(%w+)[.]?") do
        if p == nil or t == nil then
            error("Invalid match key '"..mk.."'")
        end
        t = t[p]
    end
    return t
end

table.update = function(a, b)
    for k,v in pairs(b) do
        a[k] = v
    end
end

function init(...)
    local args = {...}
    table.remove(args, 1)
    if #args == 0 then
        assert(dat_path, "dat file argument is missing")
    end

    if #args > 1 then
        match_key = table.remove(args, 1)
    end

    local dat_hash = {}
    for _, dat_path in ipairs(args) do
        local dat_file, err = io.open(dat_path, "r")
        if err then
            error("  could not open dat file '" .. dat_path .. "':" .. err)
        end

        print("  " .. dat_path)
        local objs = dat_parser(dat_lexer(dat_file, dat_path))
        dat_file:close()
        for _, obj in pairs(objs) do
            if match_key then
                local mk = get_match_key(match_key, obj)
                if mk == nil then
                    error("  missing match key '" .. match_key .. "' in one of the entries")
                end
                if dat_hash[mk] == nil then
                    dat_hash[mk] = {}
                    table.insert(dat_obj, dat_hash[mk])
                end
                table.update(dat_hash[mk], obj)
            else
                table.insert(dat_obj, obj)
            end
        end
    end
end

function get_value()
    local t = table.remove(dat_obj)
    if not t then
        return
    else
        return {
            name = t.name,
            description = t.description,
            rom_name = t.rom.name,
            size = uint(tonumber(t.rom.size)),
            users = uint(tonumber(t.users)),
            releasemonth = uint(tonumber(t.releasemonth)),
            releaseyear = uint(tonumber(t.releaseyear)),
            rumble = uint(tonumber(t.rumble)),
            analog = uint(tonumber(t.analog)),

            famitsu_rating = uint(tonumber(t.famitsu_rating)),
            edge_rating = uint(tonumber(t.edge_rating)),
            edge_issue = uint(tonumber(t.edge_issue)),
            edge_review = t.edge_review,

            enhancement_hw = t.enhancement_hw,
            barcode = t.barcode,
            esrb_rating = t.esrb_rating,
            elspa_rating = t.elspa_rating,
            pegi_rating = t.pegi_rating,
            cero_rating = t.cero_rating,
            franchise   = t.franchise,

            developer = t.developer,
            publisher = t.publisher,
            origin = t.origin,

            crc = binary(unhex(t.rom.crc)),
            md5 = binary(unhex(t.rom.md5)),
            sha1 = binary(unhex(t.rom.sha1)),
            serial = binary(t.serial or t.rom.serial),
        }
    end
end
